extensions/ql-vscode/src/run-queries.ts

```diff
@@ -718,6 +712,22 @@ const compilationFailedErrorTail = ' compilation failed. Please make sure there
   ' and the query and database use the same target language. For more details on the error, go to View > Output,' +
   ' and choose CodeQL Query Server from the dropdown.';
 
+/**
+ * Gets metadata for a query, if it exists.
+ * @param cliServer The CLI server.
+ * @param queryPath The path to the query.
+ * @returns A promise that resolves to the query metadata, if available.
+ */
+export async function getQueryMetadata(cliServer: cli.CodeQLCliServer, queryPath: string): Promise<QueryMetadata | undefined> {
```

Comment:
I have two small suggestions, but they shouldn't block this PR and perhaps can trigger wider discussion:
* Move this function to another module (file) e.g. query-metadata-utils or query-metadata-resolver or something like that. I generally have a preference for smaller, more specific/targetted modules and this one is at 700+ lines of code.
* Call this function `tryGetQueryMetadata` to indicate that it will try its best but will not blow up if it can't get the metadata. This can be a general pattern we follow.